provider/vsphere/internal/vsphereclient/createvm.go

```diff
@@ -159,87 +177,131 @@ func (c *Client) CreateVirtualMachine(
 	datastore.DatacenterPath = datacenter.InventoryPath
 	datastore.SetInventoryPath(path.Join(folders.DatastoreFolder.InventoryPath, datastoreMo.Name))
 
-	// Ensure the VMDK is present in the datastore, uploading it if it
-	// doesn't already exist.
-	resourcePool := object.NewResourcePool(c.client.Client, args.ResourcePool)
-	taskWaiter := &taskWaiter{args.Clock, args.UpdateProgress, args.UpdateProgressInterval}
-	vmdkDatastorePath, releaseVMDK, err := c.ensureVMDK(ctx, args, datastore, datacenter, taskWaiter)
+	spec, err := c.createImportSpec(ctx, args, datastore)
 	if err != nil {
-		return nil, errors.Trace(err)
+		return nil, errors.Annotate(err, "creating import spec")
 	}
-	defer releaseVMDK()
 
-	// Import the VApp, creating a temporary VM. This is necessary to
-	// import the VMDK, which exists in the datastore as a not-a-disk
-	// file type.
-	args.UpdateProgress("creating import spec")
-	importSpec, err := c.createImportSpec(ctx, args, datastore, vmdkDatastorePath)
+	vmFolder, err := c.EnsureVMFolder(ctx, vmTemplatePath(args))
 	if err != nil {
-		return nil, errors.Annotate(err, "creating import spec")
+		return nil, errors.Trace(err)
 	}
-	importSpec.ConfigSpec.Name += ".tmp"
 
-	args.UpdateProgress(fmt.Sprintf("creating VM %q", args.Name))
-	c.logger.Debugf("creating temporary VM in folder %s", vmFolder)
+	importSpec := spec.ImportSpec
+	args.UpdateProgress(fmt.Sprintf("creating template VM %q", vmTemplateName(args.Name)))
+	c.logger.Debugf("creating template VM in folder %s", vmFolder)
 	c.logger.Tracef("import spec: %s", pretty.Sprint(importSpec))
+	resourcePool := object.NewResourcePool(c.client.Client, args.ResourcePool)
 	lease, err := resourcePool.ImportVApp(ctx, importSpec, vmFolder, nil)
 	if err != nil {
-		return nil, errors.Annotatef(err, "failed to import vapp")
+		return nil, errors.Annotate(err, "failed to import vapp")
 	}
-	info, err := lease.Wait(ctx, nil)
+	info, err := lease.Wait(ctx, spec.FileItem)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
+
+	updater := lease.StartUpdater(ctx, info)
+	defer updater.Done()
+
+	ovaLocation, ovaReadCloser, err := args.ReadOVA()
+	if err != nil {
+		return nil, errors.Annotate(err, "fetching OVA")
+	}
+	defer ovaReadCloser.Close()
+	ovaTarReader := tar.NewReader(ovaReadCloser)
+	for {
+		header, err := ovaTarReader.Next()
+		if err != nil {
+			return nil, errors.Annotate(err, "reading OVA")
+		}
+		if strings.HasSuffix(header.Name, ".vmdk") {
+			item := info.Items[0]
```

Comment:
If the OVA has multiple vmdk's, how do we select the right one? It appears we just upload the first one.

Reply:
This should be filed as a new bug. We're mirroring the old business logic here. This hasn't been an issue for Juju thus far as we choose OVAs from a known source that only has a single VMDK.